plugin/src/main/java/io/jenkins/plugins/forensics/git/miner/GitRepositoryMiner.java

```diff
@@ -69,68 +78,100 @@ public RepositoryStatistics mine(final Collection<String> absoluteFileNames, fin
         }
 
-        RepositoryStatistics analyze(final Repository repository, final Collection<String> files) {
+        RepositoryStatistics analyze(final Repository repository, final Git git, final List<RevCommit> commits)
+                throws IOException {
             RepositoryStatistics statistics = new RepositoryStatistics();
-            logger.logInfo("Invoking Git miner to create statistics for all available files");
-            logger.logInfo("Git working tree = '%s'", getWorkTree(repository));
-
             FileStatisticsBuilder builder = new FileStatisticsBuilder();
-            List<FileStatistics> fileStatistics = files.stream()
-                    .map(builder::build)
-                    .map(file -> analyzeHistory(repository, file))
-                    .collect(Collectors.toList());
-            statistics.addAll(fileStatistics);
-
-            logger.logInfo("-> created statistics for %d files", statistics.size());
-
+            List<String> files = new ArrayList<>();
+            Map<String, FileStatistics> fileStatistics = new HashMap<>();
+            Set<String> filesInHead = new FilesCollector(repository).findAllFor(repository.resolve(Constants.HEAD));
+            for (int i = commits.size() - 1; i >= 0; i--) {
+                if (i == commits.size() - 1) {
+                    files = getFilesFromCommit(repository, git, null, commits.get(i).getName());
+                }
+                else {
+                    if (i > 0) {
+                        files = getFilesFromCommit(repository, git, commits.get(i + 1).getName(),
+                                commits.get(i).getName());
+                    }
+                }
+                int finalI = i;
+                files.forEach(f -> fileStatistics.computeIfAbsent(f, builder::build)
+                        .inspectCommit(commits.get(finalI).getCommitTime(), getAuthor(commits.get(finalI))));
+            }
+            fileStatistics.keySet().removeIf(f -> !filesInHead.contains(f));
+            statistics.addAll(fileStatistics.values());
             return statistics;
         }
 
-        private FileStatistics analyzeHistory(final Repository repository, final FileStatistics fileStatistics) {
-            try (Git git = new Git(repository)) {
-                Iterable<RevCommit> commits = git.log().addPath(fileStatistics.getFileName()).call();
-                commits.forEach(c -> fileStatistics.inspectCommit(c.getCommitTime(), getAuthor(c)));
-                return fileStatistics;
+        private List<String> getFilesFromCommit(final Repository repository, final Git git, final String oldCommit,
+                final String newCommit) {
+            List<String> filePaths = new ArrayList<>();
+
+            try {
+                final List<DiffEntry> diffEntries = git.diff()
+                        .setOldTree(getTreeParser(repository, oldCommit))
+                        .setNewTree(getTreeParser(repository, newCommit))
```

Comment:
What happens if we do not set `oldTree` at all? Isn't the previous commit automatically used?

Reply:
It does not seem to work without ``oldTree``, but I will take a closer look into this.